Fix: reduce huge bundle size by adding ADVANCED_OPTIMIZATIONS when compiling protobuf JS

sayjeyhi opened the pull request:
This will close: https://github.com/protocolbuffers/protobuf-javascript/issues/124

sayjeyhi commented:
I've used: https://github.com/protocolbuffers/protobuf/pull/5509 and this also fixes https://github.com/protocolbuffers/protobuf-javascript/issues/92

After optimization, the bundle size went from ~156KB to ~47KB (uncompressed)

dibenede commented:
Thanks for porting this over!

It looks like JSC is unhappy with @export annotations on functions that already have other visibility modifiers. If I'm aligning the errors correctly, some of the offending functions are currently marked as @protected (e.g. Message.getRepeatedFloatingPointField). Is @export really required for these?

sayjeyhi commented:
> Thanks for porting this over!
> 
> It looks like JSC is unhappy with @export annotations on functions that already have other visibility modifiers. If I'm aligning the errors correctly, some of the offending functions are currently marked as @Protected (e.g. Message.getRepeatedFloatingPointField). Is @export really required for these?

sayjeyhi commented:
> It looks like JSC is unhappy with @export annotations on functions that already have other visibility modifiers. If I'm aligning the errors correctly, some of the offending functions are currently marked as @Protected (e.g. Message.getRepeatedFloatingPointField). Is @export really required for these?

Yes, since we are removing them from deps, we need to tell JSC to bring the required ones by putting `@export` annotation there, to allow them to be bundled and used, those duplicate visibility modifier errors were only warnings, but I've fixed them on my last commit.
Still 2 tests are failing there:
```
Failures:
1) protoBinaryTest testUnknownExtension
  Message:
    TypeError: jspb.BinaryWriter is not a constructor
  Stack:
        at <Jasmine>
        at UserContext.<anonymous> (/Users/sayjeyhi/Desktop/Projects/github/protobuf-javascript/commonjs_out/binary/proto_test.js:643:18)
        at <Jasmine>
        at processImmediate (internal/timers.js:464:21)

2) Message test suite testCopyInto_notSameType
  Message:
    TypeError: Cannot read property 'error' of undefined
  Stack:
        at <Jasmine>
        at Object.goog.testing.asserts.raiseException (/Users/sayjeyhi/Desktop/Projects/github/protobuf-javascript/commonjs_out/test_node_modules/closure_asserts_commonjs.js:136:401)
        at goog.testing.asserts.assertThrows (/Users/sayjeyhi/Desktop/Projects/github/protobuf-javascript/commonjs_out/test_node_modules/closure_asserts_commonjs.js:97:24)
        at UserContext.<anonymous> (/Users/sayjeyhi/Desktop/Projects/github/protobuf-javascript/commonjs_out/message_test.js:567:13)
        at <Jasmine>
        at processImmediate (internal/timers.js:464:21)

87 specs, 2 failures
Finished in 0.081 seconds
Randomized with seed 81572 (jasmine --random=true --seed=81572)
```
Do you have any idea on how we can fix them?

sayjeyhi commented:
I fixed one of the tests, but the other one is still failing:
```
Failures:
1) Message test suite testCopyInto_notSameType
  Message:
    TypeError: Cannot read property 'error' of undefined
  Stack:
        at <Jasmine>
        at Object.goog.testing.asserts.raiseException (/Users/sayjeyhi/Desktop/Projects/github/protobuf-javascript/commonjs_out/test_node_modules/closure_asserts_commonjs.js:136:401)
        at goog.testing.asserts.assertThrows (/Users/sayjeyhi/Desktop/Projects/github/protobuf-javascript/commonjs_out/test_node_modules/closure_asserts_commonjs.js:97:24)
        at UserContext.<anonymous> (/Users/sayjeyhi/Desktop/Projects/github/protobuf-javascript/commonjs_out/message_test.js:567:13)
        at <Jasmine>
        at processImmediate (internal/timers.js:464:21)

87 specs, 1 failure
Finished in 0.078 seconds
Randomized with seed 58126 (jasmine --random=true --seed=58126)


[17:40:50] 'test_commonjs' errored after 213 ms
[17:40:50] Error: Command failed: sh -c cd commonjs_out && JASMINE_CONFIG_PATH=jasmine.json NODE_PATH=test_node_modules ../node_modules/.bin/jasmine
```

dibenede commented:
It looks like it's not finding goog.global.console.error in commonjs testing mode. It looks like the closure output mode tests do pass though..

Digging around, I see that we have a rewriting script to produce a new version message_test.js for the purposes of testing commonjs. I'm suspicious of that script and how it  may be interacting with tree shaking:

https://github.com/protocolbuffers/protobuf-javascript/blob/a428c58273abad07c66071d9753bc4d1289de426/commonjs/rewrite_tests_for_commonjs.js#L69

sayjeyhi commented:
@dibenede, Thanks for the review, Yeah I've checked them and even tried to require the assertion library, but it seems
The tests were passing without a reliable reason and broke with our unwanted libs removal.

cuz only this test is failing, because `assertThrows` used in this test:
```
  it('testCopyInto_notSameType', function() {
    var a = new proto.jspb.test.TestClone();
    var b = new proto.jspb.test.Simple1(['str', ['s1', 's2']]);

    var e = assertThrows(function() {
      jspb.Message.copyInto(a, b);
    });
    assertContains('should have the same type', e.message);
  });
```

and `assertThrows`  is using `raiseException` internally:

```
goog.testing.asserts.raiseException = function(comment, opt_message) {
  var e = new goog.testing.JsUnitException(comment, opt_message);

  var testCase = _getCurrentTestCase();
  if (testCase) {
    testCase.raiseAssertionException(e);
  } else {
    goog.global.console.error(
        'Failed to save thrown exception: no test case is installed.');
    throw e;
  }
};
```
when this `_getCurrentTestCase` couldn't get the test case else to block fires, and `goog.global.console` is used without checking if it is imported and it is giving us an error.

IDK, maybe we can ignore this test, or we can rewrite it using other test functions rather than `assertThrows`

dibenede commented:
I attempted rewriting as well as deleting it, but there are other errors that crop up:

```
1) Message test suite testExtendedMessageEnsureObject
  Message:
    TypeError: Cannot read properties of undefined (reading 'a_key')
  Stack:
        at <Jasmine>
        at UserContext.<anonymous> (protobuf-javascript/commonjs_out/message_test.js:742:55)
        at <Jasmine>
        at listOnTimeout (node:internal/timers:564:17)
        at process.processTimers (node:internal/timers:507:7)
```

Based off the rewrite attempt, I think it's being set off by any assert* usage :/

sayjeyhi commented:
I removed that test, and then the tests ran fine!

I fixed the test to throw an exception and then the test is running fine now without removing it.

dibenede commented on gulpfile.js:
```diff
@@ -118,7 +117,12 @@ function genproto_group3_commonjs_strict(cb) {
 
 
-function getClosureCompilerCommand(exportsFile, outputFile) {
+function getClosureCompilerCommand(exportsFile, outputFile, keepSymbols) {
```
I think one of the main things is to restore the `@protected` (and `@private`, if applicable) annotations. I think the way forward is to let the tests build in simple optimization mode. While they're currently invoking the compiler with `keepSymbols=true`, there's an underlying step that doesn't (`gen_google_protobuf_js`).

Maybe we should fixup the gulp tasks to something like:

```
function enable_advanced_optimizations(cb) {
  USE_ADVANCED_OPTIMIZATIONS = true;
  cb();
}


function disable_advanced_optimizations(cb) {
  USE_ADVANCED_OPTIMIZATIONS = false;
  cb();
}

exports.build = series(exports.build_protoc_plugin,
                      genproto_wellknowntypes,
                      gen_google_protobuf_js);

// ...

exports.test_closure = series(disable_advanced_optimizations, exports.deps,
                              test_closure);

exports.test_commonjs = series(disable_advanced_optimizations, exports.make_commonjs_out,
                               test_commonjs);
// ...

exports.dist = series(enable_advanced_optimizations, exports.build)
```

sayjeyhi replied:
But I think the test should test exactly how the code will be shipped. Exporting those methods will not change the functionality of the code. Maybe we can have an alpha minor version change(3.22.0-alpha) release to see if the bundle size and everything look fine, then release it. `230kB` minified is really too much for this library! and this release should be first priority IMO.

https://bundlephobia.com/package/google-protobuf@3.21.0

dibenede replied:
Ideally, yes, we should be able to test with optimizations off and on. No argument there. Ditto for wanting to cut down the bundle size and get a release out.

My understanding is that it is not uncommon for us to leave the tests unoptimized. While the fully `@export`'d code may be functionally ok, I think it obfuscates the real API and it would be bad if folks started relying on things outside of the contract we want to provide.

@lukesandberg Any thoughts?

sayjeyhi replied:
We really need this, guys. It took more than a month already

dibenede replied:
Luke and I discussed some things offline it seems like, as a library, we are kind of forced to `@export` more than I would otherwise like. So we'll keep going with your current approach and I'll come back later to add some appropriate warning comments to what should be internal stuff.

lukesandberg commented on message.js:
```diff
@@ -192,6 +198,7 @@ jspb.Message.GENERATE_FROM_OBJECT = goog.define(
  * @define {boolean} Whether to generate toString methods for objects. Turn
  *     this off if you do not use toString in your project and want to trim it
  *     from the compiled JS.
+ * @export
```
this shouldn't need to be exported as it is only referenced in this file iirc

sayjeyhi replied:
removed

lukesandberg commented on message.js:
```diff
@@ -228,7 +236,7 @@ jspb.Message.SUPPORTS_UINT8ARRAY_ = (typeof Uint8Array == 'function');
  * The internal data array.
  * @type {!Array}
- * @protected
+ * @export
```
i don't think this is accessed by any of the gencode, is it? we should be able to just change this to private with no need to export

sayjeyhi replied:
rolled back to `@protected`

lukesandberg commented on message.js:
```diff
@@ -212,6 +219,7 @@ jspb.Message.ASSUME_LOCAL_ARRAYS =
  *     reduces the size of serialized protos. The price is an extra iteration of
  *     the proto before serialization. This is enabled by default to be
  *     backwards compatible. Projects are advised to turn this flag always off.
+ * @export
```
i believe this is only accessed in this file and doesn't need to be exported

sayjeyhi replied:
`@export` removed

lukesandberg commented on message.js:
```diff
@@ -1791,12 +1820,15 @@ jspb.Message.cloneMessage = function(msg) {
   goog.asserts.assertInstanceof(toMessage, jspb.Message);
-  goog.asserts.assert(fromMessage.constructor == toMessage.constructor,
-      'Copy source and target message should have the same type.');
+
```
im guessing this was added because asserts are now getting eliminated in the bundle and now the tests are failing?

was that intentional or just a side effect of advance optimization

i don't think we want to eliminate asserts, at least not as part of this PR, we should separately consider that

dibenede replied:
This was a side effect of advanced optimization.

sayjeyhi replied:
The same condition is being checked with different format:
```js
  if(fromMessage.constructor !== toMessage.constructor)
      throw new Error('Copy source and target message should have the same type.');
```

lukesandberg replied:
Ok, sync'd with @dibenede on this issue and we decided that we can proceed with this PR and will just need to close out https://github.com/protocolbuffers/protobuf-javascript/issues/136 before releasing

sayjeyhi commented:
@dibenede, I fixed these comments from @lukesandberg as well, and the tests are running correctly

dibenede commented:
I'm actively working on #136, which is a blocker for release. I'm hoping we can cut a release next week.
